packages/contracts/evm-contracts/contracts/orderbook/OrderbookDex.sol

```diff
@@ -3,41 +3,59 @@
 pragma solidity ^0.8.20;
 
 import {Address} from "@openzeppelin/contracts/utils/Address.sol";
-import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {ERC1155Holder} from "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
 import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
 import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import {IInverseProjected1155} from "../token/IInverseProjected1155.sol";
 import {IOrderbookDex} from "./IOrderbookDex.sol";
 
 /// @notice Facilitates trading an asset that is living on a different app-chain.
-contract OrderbookDex is IOrderbookDex, ERC165, ReentrancyGuard {
+contract OrderbookDex is IOrderbookDex, ERC1155Holder, ReentrancyGuard {
     using Address for address payable;
 
-    mapping(address => mapping(uint256 => Order)) public orders;
-    mapping(address => uint256) public sellersOrderId;
+    IInverseProjected1155 internal asset;
+    mapping(address => mapping(uint256 => Order)) internal orders;
+    mapping(address => uint256) internal sellersOrderId;
 
     error OrderDoesNotExist(uint256 orderId);
     error InsufficientEndAmount(uint256 expectedAmount, uint256 actualAmount);
     error InvalidInput(uint256 input);
     error InvalidInputArity();
 
+    constructor(IInverseProjected1155 _asset) {
+        asset = _asset;
+    }
+
+    /// @notice Returns the address of the asset that is being traded.
+    function getAsset() public view virtual returns (address) {
+        return address(asset);
+    }
+
     /// @notice Returns the seller's current `orderId` (index that their new sell order will be mapped to).
-    function getSellerOrderId(address seller) external view returns (uint256) {
+    function getSellerOrderId(address seller) public view virtual returns (uint256) {
         return sellersOrderId[seller];
     }
 
     /// @notice Returns the Order struct information about an order identified by the combination `<seller, orderId>`.
-    function getOrder(address seller, uint256 orderId) external view returns (Order memory) {
+    function getOrder(address seller, uint256 orderId) public view virtual returns (Order memory) {
         return orders[seller][orderId];
     }
 
     /// @notice Creates a sell order with incremental seller-specific `orderId` for the specified `assetAmount` at specified `pricePerAsset`.
     /// @dev The order information is saved in a nested mapping `seller address -> orderId -> Order`.
     /// MUST emit `OrderCreated` event.
-    function createSellOrder(uint256 assetAmount, uint256 pricePerAsset) public {
+    function createSellOrder(
+        uint256 assetId,
+        uint256 assetAmount,
+        uint256 pricePerAsset
+    ) public virtual {
         if (assetAmount == 0 || pricePerAsset == 0) {
             revert InvalidInput(0);
         }
+        asset.safeTransferFrom(msg.sender, address(this), assetId, assetAmount, bytes(""));
```

Comment:
Do we want to ask for approval instead of transferring tokens here (I'm not sure which is better)?

We don't need to explicitly transfer anymore after the ERC1155 refactor I think. Previously we had to do it we needed to guarantee the orderID was generated according to a set of rules, but no longer care about orderIDs (instead, we just care about which tokenID is part of an order)

Reply:
I've thought about this, but if we are not transferring the tokens to the DEX... Imagine a scenario in which I create a sell order but then I transfer the assets away. How does the DEX know that the sell order is no longer valid? Yes, the fill function would not fill that order (since the transfer would fail), but from the front-end perspective, how do you discard that sell order? You'd have to monitor the asset contract for Transfer events? Doable, but seems kinda tedious. It also does not solve the situations in which the asset gets transferred away just before fill order transaction is mined (and would fail to fill that order), which seems unsolvable, but these situations might be improbable, I admit.
What do you think?

Reply:
Personally I think approvals are worth the better UX for sellers (after all, OpenSea works this way), but it does make integrating the front-end more complex. I'm not sure how we rank these aspects on importance.